adempiere/adempiere
# Improve Freight functionality

- Add Package as document
- Add Smart Browser for Generate Freight Orders from Package
- Add Smart Browser for Generate Sales Invoices from Freight Orders
- Add Smart Browser for Create Lines of Invoice (Vendor) from Freight
Orders
- Add Freight Info stub class
- Add Weight and Volume as standard calculation for freight engine
interface
- Add Freight Calculation Type for Freight Category
- Add Is Invoiced flag for Freight Category and Freight Order Line
- Add Default Product and charge for Freight Category
- Add Freight Rate and Freight for:
  - Sales Order
  - Distribution Order
  - Outbound Order
  - Package
- Add Freight Line reference for Invoice Line

Nice...

Here a test case using Garden World:
## Technical changes for work

### Apply changes from XML's

### Add Model validator for Freight Management
Just add model validator in your tenant based on freight management like is follow:

```Java
org.eevolution.model.FreightModelValidator
```

## Functional setup

### Create Document Types

/Create document types for:

*   Package
*   Freight Order

### Configure Default UOM:

Configure Default UOM for:

*   Weight
*   Volume
*   Length
Setup it in Client information

### Configure default invoicing rule for freight category

Freight category handle all rules of invoicing, take care of mark as true the "Is Invoiced" field

*   Set to true the "Is Invoiced" field
*   Select a Product or charge: this will be used for generate invoice for customer and create lines of vendor invoice
### Setup Shipper and Business Partner
If you want create a AP Invoice just set a business partner inside shipper configuration:

*   Create a Shipper
*   Link a Business Partner to Shipper
*   Add Freight Rules (Inside second tab of Shipper):
    *   Set a Freight amount and a route for it. If you have a question or is a very long process for setup jus use "UPS" instead and this Shipper have all done to work. If you use "UPS" shipper it have a freight rule of 20$ based on weight carried.

## General Info:
The developer test was:

### Data:

*   Document Types created:
    *   Package
    *   Freight Order
*   Shipper changes:
    *   Add Business Partner **Joe Block** to shipper **UPS**
    *   Using default freight rule of **UPS**
*   Freight Category changes
    *   Set to true **Is Invoiced** flag
    *   Set product **Azalea Bush**
    *   The freight category used for this example was **Ground**
*   Tenant changes
    *   Add Model validator for freight management
    *   Add default UOM for Weight, Volume, Length
    

### Testing Process:

1.  Create a Package:

*   Warehouse: Fertilizer
*   Shipper: UPS
*   Business Partner: GardenUser BP
*   Business Partner Location: Old York
*   Delivery Via Rule: Shipper
*   Freight Cost Rule: Calculated
*   Freight Category: Ground
*   Weight: 35
*   Result after save:
    *   Freight: UPS_Ground
    *   Freight Rate: 20
    *   Freight Amount: 700 (35 \* 20)
*   Complete Package

2.  Create Package:

*   Warehouse: Fertilizer
*   Shipper: UPS
*   Business Partner: GardenAdmin BP
*   Business Partner Location: Near Village
*   Delivery Via Rule: Shipper
*   Freight Cost rule: Calculated
*   Freight Category: Ground
*   Weighnt: 40
*   Result after save:
    *   Freight: UPS_Ground
    *   Freight Rate: 20
    *   Freight Amount: 800 (40 \* 20)
*   Complete Package

1.  Generate a Freight Order from Package: Use Smart Browser created in **Distribution Management** -> **Create Freight Order from Package**
2.  Generate Sales Invoices from Freight Order: this Smart browser generate sales invoices based on Business Partner of freight order lines. You can see this Browser in **Quote-to-Invoice** -> **Sales Invoices** -> **Generate Sales Invoice from Freight Order**
*   The product / charge to be invoiced is the defined on freight category and rounded on freight order line: **Azalea Bush**
*   The invoice quantity is the Weight / Volume of freight order line according with freight category rule used
*   The invoice price is the current price of sales price list for business partner

3.  Create Invoice (Vendor) lines from Freigh Order: This allow create lines of invoice from Freight orders, note that the AP document must have the same business partner configured on shipper **Joe Block**

*   The product / charge to be invoiced is the defined on freight category and rounded on freight order line: **Azalea Bush**
*   The quantity to invoice is the (Weight or Volume according with freight category rule)
*   If the freight category is created using charge insteda product then the quantity to invoice is **1**
*   The Invoice price is the result of **Freight Amount** / (**Weight** or **Volume** according with freight category rule)
